akka-stream/src/main/scala/akka/stream/Shape.scala

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream
+
+import scala.collection.immutable
+import scala.collection.JavaConverters._
+
+sealed abstract class InPort
+sealed abstract class OutPort
+
+final class Inlet[-T](override val toString: String) extends InPort
+final class Outlet[+T](override val toString: String) extends OutPort
+
+abstract class Shape {
+  /**
+   * Scala API: get a list of all input ports
+   */
+  def inlets: immutable.Seq[Inlet[_]]
+
+  /**
+   * Scala API: get a list of all output ports
+   */
+  def outlets: immutable.Seq[Outlet[_]]
+
+  /**
+   * Create a copy of this Ports object, returning the same type as the
+   * original; this constraint can unfortunately not be expressed in the
+   * type system.
+   */
+  def deepCopy(): Shape
+
+  /**
+   * Java API: get a list of all input ports
+   */
+  def getInlets: java.util.List[Inlet[_]] = inlets.asJava
+
+  /**
+   * Java API: get a list of all output ports
+   */
+  def getOutlets: java.util.List[Outlet[_]] = outlets.asJava
+}
+
+/**
+ * Java API for creating custom Shape types.
+ */
+abstract class AbstractShape extends Shape {
+  def allInlets: java.util.List[Inlet[_]]
+  def allOutlets: java.util.List[Outlet[_]]
+
+  final override lazy val inlets = allInlets.asScala.toList
+  final override lazy val outlets = allOutlets.asScala.toList
+
+  final override def getInlets = allInlets
+  final override def getOutlets = allOutlets
+}
+
+object EmptyShape extends Shape {
+  override val inlets = Nil
+  override val outlets = Nil
```

Comment:
in general I think it is risky to not define the type, also for overrides
e.g.

    scala> abstract class Foo { def hey: List[String]}
    defined class Foo

    scala> abstract class Bar extends Foo { override def hey = Nil}
    defined class Bar

    scala> (new Bar).hey
    res5: scala.collection.immutable.Nil.type = List()

Since this is public api it means that we could not later change `hey` in `Bar` to anything else than Nil.